core/sql/optimizer/BindRelExpr.cpp

```diff
@@ -10165,22 +10165,35 @@ base table then the old version of the row will have to be deleted from
 indexes, and a new version inserted. Upsert is being transformed to merge
 so that we can delete the old version of an updated row from the index.
 
-Upsert is also converted into merge when there are omitted cols with default values and 
-TRAF_UPSERT_WITH_INSERT_DEFAULT_SEMANTICS is set to  OFF in case of aligned format table or 
+Upsert is also converted into merge when TRAF_UPSERT_MODE is set to MERGE and 
+there are omitted cols with default values in case of aligned format table or 
 omitted current timestamp cols in case of non-aligned row format
 */
 NABoolean Insert::isUpsertThatNeedsMerge(NABoolean isAlignedRowFormat, NABoolean omittedDefaultCols,
                                    NABoolean omittedCurrentDefaultClassCols) const
 {
+  // The necessary conditions to convert upsert to merge and
   if (isUpsert() && 
       (NOT getIsTrafLoadPrep()) && 
       (NOT (getTableDesc()->isIdentityColumnGeneratedAlways() && getTableDesc()->hasIdentityColumnInClusteringKey())) && 
       (NOT (getTableDesc()->getClusteringIndex()->getNAFileSet()->hasSyskey())) && 
-       ((getTableDesc()->hasSecondaryIndexes()) ||
-         (( NOT isAlignedRowFormat) && omittedCurrentDefaultClassCols) ||
-             ((isAlignedRowFormat && omittedDefaultCols
-              && (CmpCommon::getDefault(TRAF_UPSERT_WITH_INSERT_DEFAULT_SEMANTICS) == DF_OFF)))
-       ))
+        // table has secondary indexes or
+        (getTableDesc()->hasSecondaryIndexes() ||
+          // CQD is set to MERGE  
+          ((CmpCommon::getDefault(TRAF_UPSERT_MODE) == DF_MERGE) &&
+            // omitted current default columns with non-aligned row format tables
+            // or omitted default columns with aligned row format tables 
+            (((NOT isAlignedRowFormat) && omittedCurrentDefaultClassCols) ||
+            (isAlignedRowFormat && omittedDefaultCols))) ||
+          // CQD is set to Optimal, for non-aligned row format with omitted 
+          // current columns, it is converted into merge though it is not
+          // optimal for performance - This is done to ensure that when the 
+          // CQD is set to optimal, non-aligned format would behave like 
+          // merge when any column is  omitted 
```

Comment:
What I intended to say is this:
Non-aligned format table usually behaves like merge though it is not changed into MERGE statement because Trafodion engine doesn't store the omitted default column in the raw hbase table.
When the upsert statement has omitted current default columns, it is changed into MERGE statement though it is not optimal because we want it to behave like above for this case also. Otherwise, it might be confusing to the application.